adapters/beachfront/beachfront.go

```diff
@@ -197,86 +186,91 @@ func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *a
 	return reqs, errs
 }
 
-func preprocess(request *openrtb.BidRequest) (beachfrontReqs beachfrontRequests, errs []error) {
-	var videoImps = make([]openrtb.Imp, 0)
-	var bannerImps = make([]openrtb.Imp, 0)
+func preprocess(request *openrtb.BidRequest) (beachfrontReqs requests, errs []error) {
+	var videoImps = make([]openrtb.Imp, 0, len(request.Imp))
+	var gotBanner bool
 
 	for i := 0; i < len(request.Imp); i++ {
 		if request.Imp[i].Banner != nil && ((request.Imp[i].Banner.Format[0].H != 0 && request.Imp[i].Banner.Format[0].W != 0) ||
 			(request.Imp[i].Banner.H != nil && request.Imp[i].Banner.W != nil)) {
-			bannerImps = append(bannerImps, request.Imp[i])
+			gotBanner = true
 		}
 
 		if request.Imp[i].Video != nil {
-			videoImps = append(videoImps, request.Imp[i])
+			imp := request.Imp[i]
+
+			if request.Site != nil && request.Site.Page != "" {
+				_, imp.Secure = isPageSecure(request.Site.Page)
+			} else if request.Imp[i].Secure != nil {
+				imp.Secure = request.Imp[i].Secure
+			} else {
+				imp.Secure = &zero
+			}
+
+			imp.Banner = nil
+			videoImps = append(videoImps, imp)
 		}
 	}
 
-	if len(bannerImps)+len(videoImps) == 0 {
+	if len(videoImps) == 0 && !gotBanner {
 		errs = append(errs, errors.New("no valid impressions were found in the request"))
 		return
 	}
 
-	if len(bannerImps) > 0 {
-		request.Imp = bannerImps
+	if gotBanner {
```

Comment:
It is possible to have errors from getBannerRequest and still have a valid banner, so I don't see any appropriate action that I could  take _exactly_ here, but I found several related optimizations. If all banners errored out (failed on ext or appId), the Slots will be empty, so I'm now checking that in getBannerRequest before adding anything else (ip, secure. etc), so in that case I now skip a bunch of unnecessary processing on an empty banner request.

I realized that how I had it, if I did get banner errors, they would get overwritten by the errs returned from getVideoRequests, even if the later is nil. That's no good. Previously (current master) I had merged []error responses.  I changed that to pass a []error pointer to preprocess and down the line. That creates a fairly big cascade of changes, but I think it is much better, not just because it avoids that overwrite.

I removed the test in prepVideoRequestExt to verify a video element, as it is impossible for anything that is not a video imp to get into videoImps (L211), so that was just a waste of a few cpu cycles.

Reply:
@muncha Sorry but I am not sure I follow why you needed to pass the `errs` slice to `preprocess` rather than having it send back an error slice

Reply:
Right. Duh. I was looking for places that I could use pointers, but here there is really no point, as it is nil when I send it in.

In my current local changes, I've eliminated passing the []error in and have preprocess back to returning it. In preprocess, I am creating the slice and passing a pointer to that []error to getBannerRequest and getVideoRequests.